File: runme/mcgen/local/datan/solsrc/s4min.f
      PROGRAM S4MIN
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(NR=5,NPL=1000,T0=-10.25D0,ONE=1.D0)
      COMMON /COM/ X(NR),RANDOM(1000),XPL(1000),YPL(1000),
     +CX(NR,NR),SCRAT(NR,NR+1),XMC(NR),SCRAT1(NR,NR),LIST(NR)
C COMMON /COM/ is needed for Microsoft FORTRAN compiler
      COMMON /CMINBH/ T(100),HIST(100),DELTAT,NT,NEVENT
      DOUBLE PRECISION MINBLP,MINBSQ
      CHARACTER*75 CAPT,TX,TY
      CHARACTER*40 TEXT(NR)
      LOGICAL OK
      EXTERNAL MINBLP,MINBSQ
      DATA TEXT /'fraction of events in 1st BW function   ',
     +'mean of 1st BW function  (>= -5., <=5.) ',
     +'FWHM of 1st BW function  (>0.)          ',
     +'mean of 2nd BW function  (>= -5., <=5.) ',
     +'FWHM of 2nd BW function  (>0.)          '/
C identify program to user
      WRITE(*,*)' program S4MIN performs fit of 2 Breit-Wigner'
      WRITE(*,*)' functions to sample or histogram'
      WRITE(*,*)' by minimization'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter number of events (<= 1000)'
      WRITE(*,*)' >'
      READ(*,*)NEVENT
      WRITE(*,*)' Enter number of histogram bins (<= 100)'
      WRITE(*,*)' >'
      READ(*,*)NT
      DO 1 I=1,NR
        WRITE(*,*)' Enter ',TEXT(I)
        WRITE(*,*)' >'
        READ(*,*)XMC(I)
        WRITE(*,*)' Enter initial value for that parameter'
        WRITE(*,*)' >'
        READ(*,*)X(I)
        WRITE(*,*)' Enter'
        WRITE(*,*)' 1 - if parameter is to be kept variable'
        WRITE(*,*)' 0 - if parameter is to fixed'
        WRITE(*,*)' >'
        READ(*,*)LIST(I)
1     CONTINUE
      WRITE(*,*)' Choose'
      WRITE(*,*)' 1 - likelihood minimization with Poisson stat.'
      WRITE(*,*)' 2 - minimization of sum of squares'
      WRITE(*,*)' >'
      READ(*,*)ICH
C draw sample
      NSEED1=87654
      NSEED2=98765
      CALL RNE2IN(NSEED1,NSEED2)
      CALL RNECUY(RANDOM,NEVENT)
      DO 2 I=1,NEVENT
        IF(RANDOM(I).LT.XMC(1)) THEN
          CALL RNBW(XMC(2),XMC(3),RANDOM(I),1)
        ELSE
          CALL RNBW(XMC(4),XMC(5),RANDOM(I),1)
        END IF
2     CONTINUE
C prepare and fill histogram
      DELTAT=2.D0*ABS(T0)/DBLE(NT)
      DO 5 I=1,NT
        T(I)=T0+DBLE(I-0.5D0)*DELTAT
5     CONTINUE
      CALL SMHSIN(HIST,T0,DELTAT,NT)
      DO 10 I=1,NEVENT
        CALL SMHSFL(HIST,T0,DELTAT,NT,RANDOM(I),1.D0)
10    CONTINUE
C set NRED
      NRED=0
      DO 15 I=1,NR
        IF(LIST(I).EQ.1) NRED=NRED+1
15    CONTINUE
C minimize with MINSIM
      NSTEP=0
      EPSILN=0.D0
      WRITE(*,*)' minimization with MINSIM'
      WRITE(*,'(3(A,I4),A,5I2)')' N = ',NT,', NR = ',NR,', NRED = ' ,
     +NRED,', LIST = ',LIST
      WRITE(*,'(A,5F10.5)')' first approx.: X = ',X
      IF(ICH.EQ.1) THEN
        CALL MINSIM(X,NR,NRED,LIST,MINBLP,FMIN,EPSILN,NSTEP,SCRAT)
      ELSE
        CALL MINSIM(X,NR,NRED,LIST,MINBSQ,FMIN,EPSILN,NSTEP,SCRAT)
      END IF
      IF(NSTEP.LT.0) THEN
        WRITE(*,*)' minimization procedure failed'
        STOP
      END IF
      WRITE(*,'(A,F8.2,A,I6,/,A,(5F10.5))')
     +' result of minimization: FMIN = ', FMIN,', NSTEP =',NSTEP,
     +' X =',X
C determine covariance matrix
      FACT=DBLE(1.)
      IF(ICH.EQ.1) THEN
        CALL MINCOV(X,NR,NRED,LIST,MINBLP,FACT,SCRAT,SCRAT1,CX,OK)
      ELSE
        CALL MINCOV(X,NR,NRED,LIST,MINBSQ,FACT,SCRAT,SCRAT1,CX,OK)
      END IF
      IF(.NOT.OK) THEN
        WRITE(*,*)' determination of covariance matrix fails'
        STOP
      END IF
      WRITE(*,*)' covariance matrix CX = '
      CALL MTXWRT(CX,NRED,NRED)
      WRITE(*,'(/)')
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
C plot
C prepare caption
      CAPT(1:38)='x_1#=******, x_2#=******, x_3#=******,'
      CAPT(39:63)=' x_4#=******, x_5#=******'
      WRITE(CAPT(6:11),'(F6.3)') X(1)
      WRITE(CAPT(19:24),'(F6.3)') X(2)
      WRITE(CAPT(32:37),'(F6.3)') X(3)
      WRITE(CAPT(45:50),'(F6.3)') X(4)
      WRITE(CAPT(58:63),'(F6.3)') X(5)
      LCAPT=63
      TX='y'
      LTX=1
      TY='N(y)'
      LTY=4
C draw curve corresponding to solution
      FACT=DBLE(NEVENT)*DELTAT
      DPL=2.D0*ABS(T0)/DBLE(NPL-1)
      DO 30 IPL=1,NPL
        XPL(IPL)=T0+DBLE(IPL-1)*DPL
        YPL(IPL)=FACT*(X(1)*BREWIG(XPL(IPL),X(2),X(3))+(ONE-X(1))
     +  *BREWIG(XPL(IPL),X(4),X(5)))
30    CONTINUE
      CALL GRHSCV(XPL,YPL,NPL,
     *HIST,T0,DELTAT,NT,TX,LTX,TY,LTY,CAPT,LCAPT,NWS)
      END
C ------------------------------------------------------------------
      SUBROUTINE RNBW(A,GAMMA,R,N)
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(PI=3.14159D0,HALF=0.5D0)
      DIMENSION R(N)
      CALL RNECUY(R,N)
      DO 10 I=1,N
        R(I)=A+HALF*GAMMA*TAN(PI*(R(I)-HALF))
10    CONTINUE
      END
C ------------------------------------------------------------------
      DOUBLE PRECISION FUNCTION MINBLP(X,N)
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      COMMON /CMINBH/ T(100),HIST(100),DELTAT,NT,NEVENT
      DIMENSION X(N)
      PARAMETER(ZERO=0.D0,ONE=1.D0)
      FNORM=DBLE(NEVENT)*DELTAT
      MINBLP=ZERO
      DO 10 I=1,NT
C GI is the value of the probability density of the population at T(I)
C (by replacing the RHS of the following statement it can be
C  changed from normal to any desired distribution)
        GI=X(1)*BREWIG(T(I),X(2),X(3))+(ONE-X(1))*BREWIG(T(I),X(4),X(5))
C normalize to number of events in sample
        GI=FNORM*GI
        ALAM=GI
        IF(ALAM.GT.ZERO) THEN
          NI=NINT(HIST(I))
          ALNLAM=LOG(ALAM)
          ALNNI=GLNGAM(DBLE(NI+1))
          MINBLP=MINBLP+ALNNI-HIST(I)*ALNLAM+ALAM
        END IF
10    CONTINUE
      END
C ------------------------------------------------------------------
      DOUBLE PRECISION FUNCTION MINBSQ(X,N)
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      COMMON /CMINBH/ T(100),HIST(100),DELTAT,NT,NEVENT
      DIMENSION X(N)
      PARAMETER(ZERO=0.D0,ONE=1.D0)
      FNORM=DBLE(NEVENT)*DELTAT
      MINBSQ=ZERO
      DO 10 I=1,NT
C GI is the value of the probability density of the population at T(I)
C (by replacing the RHS of the following statement it can be
C  changed from normal to any desired distribution)
        GI=X(1)*BREWIG(T(I),X(2),X(3))+(ONE-X(1))*BREWIG(T(I),X(4),X(5))
C normalize to number of events in sample
        GI=FNORM*GI
        IF(HIST(I).GT.ZERO) THEN
          MINBSQ=MINBSQ+(HIST(I)-GI)**2/HIST(I)
        END IF
10    CONTINUE
      END
C ------------------------------------------------------------------
      DOUBLE PRECISION FUNCTION BREWIG(X,A,GAMMA)
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
C yields Breit-Wigner function at point X with mean A and FWHM GAMMA
      PARAMETER(TOP=.6366198D0)
      BREWIG=TOP*ABS(GAMMA)/(4.D0*(X-A)**2+GAMMA**2)
      END

File: runme/mcgen/local/datan/solsrc/s4min.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

extern void LIBFUNC rnbw(double const P_T *a, double const P_T *gamma,
                         double P_T *r, integer const P_T *n);
extern double LIBFUNC minblp(double const P_T *x, integer const P_T *n);
extern double LIBFUNC minbsq(double const P_T *x, integer const P_T *n);
extern double LIBFUNC brewig(double const P_T *x, double const P_T *a,
                             double const P_T *gamma);

/* Common Block Declarations */
/* COMMON /COM/ is needed for Microsoft FORTRAN compiler */
struct {
    double x[5], random[1000], xpl[1000], ypl[1000], cx[25] /* was [5][5] */,
           scrat[30] /* was [5][6] */, xmc[5], scrat1[25] /* was [5][5] */;
    integer list[5];
} com;

struct cminbh {
    double t[100], hist[100], deltat;
    integer nt, nevent;
} LIBDATA cminbh;

/* Table of constant values */

static integer c__1 = 1, c__5 = 5, c__1000 = 1000;
static double c_dm10p25 = -10.25, c_d1 = 1.;

main(void)
{
    /* Local variables */
    static integer i, lcapt, nstep, nseed1, nseed2, ich, ipl, ltx, lty, nws,
                   nred;
    static double dpl, epsiln, fact, fmin;
    static char capt[75], tx[75], ty[75], *text[5] =
                { "fraction of events in 1st BW function",
                  "mean of 1st BW function (>= -5., <= 5.)",
                  "FWHM of 1st BW function  (>0.)",
                  "mean of 2nd BW function (>= -5., <= 5.)",
                  "FWHM of 2nd BW function  (>0.)" };
    static logical ok;
    /* System generated locals */
    integer i__1;
/* identify program to user */
    printf("%s\n", "Program S4MIN performs fit of 2 Breit-Wigner");
    printf("%s\n", "functions to sample or histogram");
    printf("%s\n\n", "by minimization.");
/* ask for input */
    printf("%s\n", "Enter number of events (<= 1000):");
    printf("%s", "> ");
    scanf("%li", &cminbh.nevent);
    printf("%s\n", "Enter number of histogram bins (<= 100):");
    printf("%s", "> ");
    scanf("%li", &cminbh.nt);
    for (i = 1; i <= 5; ++i) {
        printf("%s%s:\n", "Enter ", text[i-1]);
        printf("%s", "> ");
        scanf("%lf", &com.xmc[i-1]);
        printf("%s\n", "Enter initial value for that parameter:");
        printf("%s", "> ");
        scanf("%lf", &com.x[i-1]);
        printf("%s\n", "Enter:");
        printf("%s\n", "1 - if parameter is to be kept variable");
        printf("%s\n", "0 - if parameter is to fixed");
        printf("%s", "> ");
        scanf("%li", &com.list[i-1]);
    }
    printf("%s\n", "Choose:");
    printf("%s\n", "1 - likelihood minimization with Poisson stat.");
    printf("%s\n", "2 - minimization of sum of squares");
    printf("%s", "> ");
    scanf("%li", &ich);
/* draw sample */
    nseed1 = 87654;
    nseed2 = 98765;
    rne2in_(&nseed1, &nseed2);
    rnecuy_(com.random, &cminbh.nevent);
    i__1 = cminbh.nevent;
    for (i = 1; i <= i__1; ++i) {
        if (com.random[i - 1] < com.xmc[0]) {
            rnbw(&com.xmc[1], &com.xmc[2], &com.random[i - 1], &c__1);
        } else {
            rnbw(&com.xmc[3], &com.xmc[4], &com.random[i - 1], &c__1);
        }
    }
/* prepare and fill histogram */
    cminbh.deltat = 20.5 / (double) cminbh.nt;
    i__1 = cminbh.nt;
    for (i = 1; i <= i__1; ++i) {
        cminbh.t[i - 1] = (i - .5) * cminbh.deltat - 10.25;
    }
    smhsin_(cminbh.hist, &c_dm10p25, &cminbh.deltat, &cminbh.nt);
    i__1 = cminbh.nevent;
    for (i = 1; i <= i__1; ++i) {
        smhsfl_(cminbh.hist, &c_dm10p25, &cminbh.deltat, &cminbh.nt,
               &com.random[i - 1], &c_d1);
    }
/* set NRED */
    nred = 0;
    for (i = 1; i <= 5; ++i) {
        if (com.list[i - 1] == 1) {
            ++nred;
        }
    }
/* minimize with MINSIM */
    nstep = 0;
    epsiln = 0.;
    printf("%s\n", "minimization with MINSIM");
    printf("%s%4li%s%4li%s%4li%s%2li%2li%2li%2li%2li\n", "N = ", cminbh.nt,
           ", NR = ", c__5, ", NRED = ", nred, ", LIST = ",
           com.list[0], com.list[1], com.list[2], com.list[3], com.list[4]);
    printf("%s%10.5lf%10.5lf%10.5lf%10.5lf%10.5lf\n", "first approx.: X = ",
           com.x[0], com.x[1], com.x[2], com.x[3], com.x[4]);
    if (ich == 1) {
        minsim_(com.x, &c__5, &nred, com.list, minblp, &fmin,
                &epsiln, &nstep, com.scrat);
    } else {
        minsim_(com.x, &c__5, &nred, com.list, minbsq, &fmin,
               &epsiln, &nstep, com.scrat);
    }
    if (nstep < 0) {
        printf("%s\n", "minimization procedure failed");
        exit(0);
    }
    printf("%s%8.2lf%s%6li\n%s%10.5lf%10.5lf%10.5lf%10.5lf%10.5lf\n",
           "result of minimization: FMIN = ", fmin, ", NSTEP =", nstep,
           "X =", com.x[0], com.x[1], com.x[2], com.x[3], com.x[4]);
/* determine covariance matrix */
    fact = 1.;
    if (ich == 1) {
        mincov_(com.x, &c__5, &nred, com.list, minblp, &fact,
                com.scrat, com.scrat1, com.cx, &ok);
    } else {
        mincov_(com.x, &c__5, &nred, com.list, minbsq, &fact,
                com.scrat, com.scrat1, com.cx, &ok);
    }
    if (! ok) {
        printf("%s\n", "determination of covariance matrix fails");
        exit(0);
    }
    printf("%s\n", "covariance matrix CX = ");
    mtxwrt_(com.cx, &nred, &nred);
    printf("\n");
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
/* plot */
/* prepare caption */
    sprintf(capt, "x_1#=%6.3lf, x_2#=%6.3lf, x_3#=%6.3lf,"
            " x_4#=%6.3lf, x_5#=%6.3lf", com.x[0], com.x[1],
            com.x[2], com.x[3], com.x[4]);
    lcapt = strlen(capt);
    memset(tx, ' ', 75);
    memset(ty, ' ', 75);
    strncpy(tx, "y", 1);
    ltx = 1;
    strncpy(ty, "N(y)", 4);
    lty = 4;
/* draw curve corresponding to solution */
    fact = (double) cminbh.nevent * cminbh.deltat;
    dpl = .020520520520520519;
    for (ipl = 1; ipl <= 1000; ++ipl) {
        com.xpl[ipl - 1] = (double) (ipl - 1) * dpl - 10.25;
        com.ypl[ipl - 1] = fact * (com.x[0] * brewig(&com.xpl[ipl - 1],
                 &com.x[1], &com.x[2]) + (1. - com.x[0]) * brewig(&
                com.xpl[ipl - 1], &com.x[3], &com.x[4]));
    }
    (void)getchar();
    grhscv_(com.xpl, com.ypl, &c__1000, cminbh.hist, &c_dm10p25, &cminbh.deltat,
           &cminbh.nt, tx, &ltx, ty, &lty, capt, &lcapt, &nws);
    return 0;
} /* main */

/* ------------------------------------------------------------------ */
void LIBFUNC rnbw(double const P_T *a, double const P_T *gamma,
                  double P_T *r, integer const P_T *n)
{
    /* System generated locals */
    integer i__1;

    /* Local variables */
    static integer i;

    /* Parameter adjustments */
    --r;

    /* Function Body */
    rnecuy_(&r[1], n);
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
        r[i] = *a + *gamma * .5 * tan((r[i] - .5) * 3.14159);
    }
    return;
} /* rnbw */

/* ------------------------------------------------------------------ */
double LIBFUNC minblp(double const P_T *x, integer const P_T *n)
{
    /* System generated locals */
    integer i__1;
    double ret_val, d__1;

    /* Local variables */
    static integer i, ni;
    static double alam, alnlam, alnni, fnorm, gi;

    /* Parameter adjustments */
    --x;

    /* Function Body */
    fnorm = (double) cminbh.nevent * cminbh.deltat;
    ret_val = 0.;
    i__1 = cminbh.nt;
    for (i = 1; i <= i__1; ++i) {
/* GI is the value of the probability density of the population at T(I) */
/* (by replacing the RHS of the following statement it can be */
/*  changed from normal to any desired distribution) */
        gi = x[1] * brewig(&cminbh.t[i - 1], &x[2], &x[3]) + (1. - x[1]) *
                brewig(&cminbh.t[i - 1], &x[4], &x[5]);
/* normalize to number of events in sample */
        gi = fnorm * gi;
        alam = gi;
        if (alam > 0.) {
            ni = nint(cminbh.hist[i - 1]);
            alnlam = log(alam);
            d__1 = (double) (ni + 1);
            alnni = glngam_(&d__1);
            ret_val = ret_val + alnni - cminbh.hist[i - 1] * alnlam + alam;
        }
    }
    return ret_val;
} /* minblp */

/* ------------------------------------------------------------------ */
double LIBFUNC minbsq(double const P_T *x, integer const P_T *n)
{
    /* System generated locals */
    integer i__1;
    double ret_val, d__1;

    /* Local variables */
    static integer i;
    static double fnorm, gi;

    /* Parameter adjustments */
    --x;

    /* Function Body */
    fnorm = (double) cminbh.nevent * cminbh.deltat;
    ret_val = 0.;
    i__1 = cminbh.nt;
    for (i = 1; i <= i__1; ++i) {
/* GI is the value of the probability density of the population at T(I) */
/* (by replacing the RHS of the following statement it can be */
/*  changed from normal to any desired distribution) */
        gi = x[1] * brewig(&cminbh.t[i - 1], &x[2], &x[3]) + (1. - x[1]) *
                brewig(&cminbh.t[i - 1], &x[4], &x[5]);
/* normalize to number of events in sample */
        gi = fnorm * gi;
        if (cminbh.hist[i - 1] > 0.) {
/* Computing 2nd power */
            d__1 = cminbh.hist[i - 1] - gi;
            ret_val += d__1 * d__1 / cminbh.hist[i - 1];
        }
    }
    return ret_val;
} /* minbsq */

/* ------------------------------------------------------------------ */
double LIBFUNC brewig(double const P_T *x, double const P_T *a,
                      double const P_T *gamma)
{
    /* System generated locals */
    double ret_val, d__1, d__2;

/* yields Breit-Wigner function at point X with mean A and FWHM GAMMA */
/* Computing 2nd power */
    d__1 = *x - *a;
/* Computing 2nd power */
    d__2 = *gamma;
    ret_val = abs(*gamma) * .6366198 / (d__1 * d__1 * 4. + d__2 * d__2);
    return ret_val;
} /* brewig */

